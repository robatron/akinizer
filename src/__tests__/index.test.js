const gulp = require('gulp');
const {
    ACTIONS,
    definePhase,
    defineRoot,
    createTaskTree,
} = require('../index');

jest.mock('gulp');

/**
 * defaultTestPhaseTreeDef is a definition of a phase tree exemplifying
 * thefollowing features:
 *
 * Phase actions:
 *  - INSTALL
 *  - VERIFY
 *  - RUN_PHASES
 *
 * Phase async:
 *  - parallel
 *  - series (default)
 */
const taskTreeRoot = defineRoot([
    definePhase('installPhase', ACTIONS.INSTALL, ['alpha', 'bravo', 'charlie']),
    definePhase('runPhase', ACTIONS.RUN_PHASES, [
        definePhase('subInstallPhase', ACTIONS.INSTALL, [
            'delta',
            'echo',
            'foxtrot',
        ]),
        definePhase('subRunPhase', ACTIONS.RUN_PHASES, [
            definePhase('subSubInstallPhase', ACTIONS.INSTALL, [
                'golf',
                'hotel',
                'india',
            ]),
            definePhase('subSubVerifyPhase', ACTIONS.VERIFY, [
                'juliett',
                'kelo',
                'lima',
            ]),
        ]),
        definePhase('subVerifyPhase', ACTIONS.VERIFY, [
            'mike',
            'november',
            'oscar',
        ]),
    ]),
    definePhase(
        'verifyPhase',
        ACTIONS.VERIFY,
        ['papa', 'quebec', 'romeo', 'sierra'],
        {
            parallel: true,
        },
    ),
]);

describe('createTaskTree', () => {
    beforeEach(() => {
        ['parallel', 'series'].forEach((asyncType) => {
            gulp[asyncType] = jest.fn((tasks) => {
                return {
                    asyncType,
                    displayName: null,
                    tasks: tasks.map((task) =>
                        typeof task === 'function' ? 'task-fn' : task,
                    ),
                };
            });
        });
    });

    it('builds a task tree from definition', () => {
        const testExports = {};
        createTaskTree(taskTreeRoot, testExports);
        expect(testExports).toMatchSnapshot();
    });

    it('exposes all tasks globally', () => {
        const testExports = {};
        createTaskTree(taskTreeRoot, testExports);

        [
            'default',
            'installPhase:alpha',
            'installPhase:bravo',
            'installPhase:charlie',
            'installPhase',
            'runPhase:subInstallPhase:delta',
            'runPhase:subInstallPhase:echo',
            'runPhase:subInstallPhase:foxtrot',
            'runPhase:subInstallPhase',
            'runPhase:subRunPhase:subSubInstallPhase:golf',
            'runPhase:subRunPhase:subSubInstallPhase:hotel',
            'runPhase:subRunPhase:subSubInstallPhase:india',
            'runPhase:subRunPhase:subSubInstallPhase',
            'runPhase:subRunPhase:subSubVerifyPhase:juliett',
            'runPhase:subRunPhase:subSubVerifyPhase:kelo',
            'runPhase:subRunPhase:subSubVerifyPhase:lima',
            'runPhase:subRunPhase:subSubVerifyPhase',
            'runPhase:subRunPhase',
            'runPhase:subVerifyPhase:mike',
            'runPhase:subVerifyPhase:november',
            'runPhase:subVerifyPhase:oscar',
            'runPhase:subVerifyPhase',
            'runPhase',
            'verifyPhase:papa',
            'verifyPhase:quebec',
            'verifyPhase:romeo',
            'verifyPhase:sierra',
            'verifyPhase',
        ].forEach((taskName) => {
            expect(testExports).toHaveProperty(taskName);
        });
    });
});
