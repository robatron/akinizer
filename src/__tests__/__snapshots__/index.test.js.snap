// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createTaskTree builds a task tree from definition 1`] = `
Object {
  "default": Object {
    "asyncType": "series",
    "displayName": "default",
    "tasks": Array [
      Object {
        "asyncType": "series",
        "displayName": "installPhase",
        "tasks": Array [
          "task-fn",
          "task-fn",
          "task-fn",
        ],
      },
      Object {
        "asyncType": "series",
        "displayName": "runPhase",
        "tasks": Array [
          Object {
            "asyncType": "series",
            "displayName": "runPhase:subInstallPhase",
            "tasks": Array [
              "task-fn",
              "task-fn",
              "task-fn",
            ],
          },
          Object {
            "asyncType": "series",
            "displayName": "runPhase:subRunPhase",
            "tasks": Array [
              Object {
                "asyncType": "series",
                "displayName": "runPhase:subRunPhase:subSubInstallPhase",
                "tasks": Array [
                  "task-fn",
                  "task-fn",
                  "task-fn",
                ],
              },
              Object {
                "asyncType": "series",
                "displayName": "runPhase:subRunPhase:subSubVerifyPhase",
                "tasks": Array [
                  "task-fn",
                  "task-fn",
                  "task-fn",
                ],
              },
            ],
          },
          Object {
            "asyncType": "series",
            "displayName": "runPhase:subVerifyPhase",
            "tasks": Array [
              "task-fn",
              "task-fn",
              "task-fn",
            ],
          },
        ],
      },
      Object {
        "asyncType": "parallel",
        "displayName": "verifyPhase",
        "tasks": Array [
          "task-fn",
          "task-fn",
          "task-fn",
          "task-fn",
        ],
      },
    ],
  },
  "installPhase": Object {
    "asyncType": "series",
    "displayName": "installPhase",
    "tasks": Array [
      "task-fn",
      "task-fn",
      "task-fn",
    ],
  },
  "installPhase:alpha": [Function],
  "installPhase:bravo": [Function],
  "installPhase:charlie": [Function],
  "runPhase": Object {
    "asyncType": "series",
    "displayName": "runPhase",
    "tasks": Array [
      Object {
        "asyncType": "series",
        "displayName": "runPhase:subInstallPhase",
        "tasks": Array [
          "task-fn",
          "task-fn",
          "task-fn",
        ],
      },
      Object {
        "asyncType": "series",
        "displayName": "runPhase:subRunPhase",
        "tasks": Array [
          Object {
            "asyncType": "series",
            "displayName": "runPhase:subRunPhase:subSubInstallPhase",
            "tasks": Array [
              "task-fn",
              "task-fn",
              "task-fn",
            ],
          },
          Object {
            "asyncType": "series",
            "displayName": "runPhase:subRunPhase:subSubVerifyPhase",
            "tasks": Array [
              "task-fn",
              "task-fn",
              "task-fn",
            ],
          },
        ],
      },
      Object {
        "asyncType": "series",
        "displayName": "runPhase:subVerifyPhase",
        "tasks": Array [
          "task-fn",
          "task-fn",
          "task-fn",
        ],
      },
    ],
  },
  "runPhase:subInstallPhase": Object {
    "asyncType": "series",
    "displayName": "runPhase:subInstallPhase",
    "tasks": Array [
      "task-fn",
      "task-fn",
      "task-fn",
    ],
  },
  "runPhase:subInstallPhase:delta": [Function],
  "runPhase:subInstallPhase:echo": [Function],
  "runPhase:subInstallPhase:foxtrot": [Function],
  "runPhase:subRunPhase": Object {
    "asyncType": "series",
    "displayName": "runPhase:subRunPhase",
    "tasks": Array [
      Object {
        "asyncType": "series",
        "displayName": "runPhase:subRunPhase:subSubInstallPhase",
        "tasks": Array [
          "task-fn",
          "task-fn",
          "task-fn",
        ],
      },
      Object {
        "asyncType": "series",
        "displayName": "runPhase:subRunPhase:subSubVerifyPhase",
        "tasks": Array [
          "task-fn",
          "task-fn",
          "task-fn",
        ],
      },
    ],
  },
  "runPhase:subRunPhase:subSubInstallPhase": Object {
    "asyncType": "series",
    "displayName": "runPhase:subRunPhase:subSubInstallPhase",
    "tasks": Array [
      "task-fn",
      "task-fn",
      "task-fn",
    ],
  },
  "runPhase:subRunPhase:subSubInstallPhase:golf": [Function],
  "runPhase:subRunPhase:subSubInstallPhase:hotel": [Function],
  "runPhase:subRunPhase:subSubInstallPhase:india": [Function],
  "runPhase:subRunPhase:subSubVerifyPhase": Object {
    "asyncType": "series",
    "displayName": "runPhase:subRunPhase:subSubVerifyPhase",
    "tasks": Array [
      "task-fn",
      "task-fn",
      "task-fn",
    ],
  },
  "runPhase:subRunPhase:subSubVerifyPhase:juliett": [Function],
  "runPhase:subRunPhase:subSubVerifyPhase:kelo": [Function],
  "runPhase:subRunPhase:subSubVerifyPhase:lima": [Function],
  "runPhase:subVerifyPhase": Object {
    "asyncType": "series",
    "displayName": "runPhase:subVerifyPhase",
    "tasks": Array [
      "task-fn",
      "task-fn",
      "task-fn",
    ],
  },
  "runPhase:subVerifyPhase:mike": [Function],
  "runPhase:subVerifyPhase:november": [Function],
  "runPhase:subVerifyPhase:oscar": [Function],
  "verifyPhase": Object {
    "asyncType": "parallel",
    "displayName": "verifyPhase",
    "tasks": Array [
      "task-fn",
      "task-fn",
      "task-fn",
      "task-fn",
    ],
  },
  "verifyPhase:papa": [Function],
  "verifyPhase:quebec": [Function],
  "verifyPhase:romeo": [Function],
  "verifyPhase:sierra": [Function],
}
`;
